### PR TITLE
[MSTR-365] feat : 문제 이의 제기 추가

### PR DESCRIPTION
**Pull request by kshired**
### Issue Number

close: #MSTR-365

### 작업 내역

- [x] 문제 이의 제기 추가
- [x] 이의 제기 추가에 따른 테스트 및 문서화 추가

### 변경사항

- 의존성 목록
  - spring-boot-starter-validation 추가

### 작업 유형

- [x] 신규 기능 추가
- [ ] 버그 수정
- [ ] 리펙토링
- [x] 문서 업데이트

### PR 특이 사항

- 슬랙 알림 보내는건 천천히 해보겠습니다. 일단 redash로 대체해도 될 듯.

### CODE REVIEW

File: src/main/kotlin/io/csbroker/apiserver/controller/v2/ProblemControllerV2.kt [hunk 1]
```diff
@@ -24,4 +30,14 @@ class ProblemControllerV2(
 
         return ApiResponse.success(findProblemDetail)
     }
+
+    @PostMapping("/{id}/challenge")
+    fun challengeProblem(
+        @LoginUser loginUser: User,
+        @PathVariable("id") id: Long,
+        @Valid @RequestBody challengeCreateRequest: ChallengeCreateRequest,
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/controller/v2/ProblemControllerV2.kt:38]: 🚫 **[ktlint]** <annotation> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Multiple annotations should not be placed on the same line as the annotated construct

File: src/main/kotlin/io/csbroker/apiserver/service/ProblemServiceImpl.kt [hunk 2]
```diff
@@ -533,4 +536,21 @@ class ProblemServiceImpl(
...
+            ?: throw EntityNotFoundException("$email 을 가진 유저는 존재하지 않습니다.")
+
+        val problem = problemRepository.findByIdOrNull(problemId)
+            ?: throw EntityNotFoundException("${problemId}번 문제는 존재하지 않는 문제입니다.")
+
+        challengeRepository.save(
+            Challenge(
+                user = user,
+                content = content,
+                problem = problem
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/service/ProblemServiceImpl.kt:552]: 🚫 **[ktlint]** <trailing-comma-on-call-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/main/kotlin/io/csbroker/apiserver/service/ProblemServiceImpl.kt [hunk 3]
```diff
@@ -533,4 +536,21 @@ class ProblemServiceImpl(
...
+            )
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/service/ProblemServiceImpl.kt:553]: 🚫 **[ktlint]** <trailing-comma-on-call-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/test/kotlin/io/csbroker/apiserver/e2e/ProblemControllerV2Test.kt [hunk 4]
```diff
@@ -122,4 +137,66 @@ class ProblemControllerV2Test {
...
+    @Test
+    @Order(2)
+    fun `문제 이의 제기`() {
+        // given
+        val shortProblem = ShortProblem(
+            title = "test11",
+            description = "test",
+            creator = adminUser!!,
+            answer = "test",
+            score = 5.0
```
Comment (kshired) [inline src/test/kotlin/io/csbroker/apiserver/e2e/ProblemControllerV2Test.kt:150]: 🚫 **[ktlint]** <trailing-comma-on-call-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/test/kotlin/io/csbroker/apiserver/e2e/ProblemControllerV2Test.kt [hunk 5]
```diff
@@ -122,4 +137,66 @@ class ProblemControllerV2Test {
...
+        )
+
+        problemRepository.save(shortProblem)
+
+        val accessToken = tokenProvider.createAuthToken(
+            email = "test50@test.com",
+            expiry = Date(Date().time + 600000),
+            role = Role.ROLE_USER.code
```
Comment (kshired) [inline src/test/kotlin/io/csbroker/apiserver/e2e/ProblemControllerV2Test.kt:158]: 🚫 **[ktlint]** <trailing-comma-on-call-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/test/kotlin/io/csbroker/apiserver/e2e/ProblemControllerV2Test.kt [hunk 6]
```diff
@@ -122,4 +137,66 @@ class ProblemControllerV2Test {
...
+        val challengeCreateRequestDto = ChallengeCreateRequest("이것은 이의제기를 위한 내용입니다.")
+        val challengeCreateRequestDtoString = objectMapper.writeValueAsString(challengeCreateRequestDto)
+
+        // when
+        val result = mockMvc.perform(
+            RestDocumentationRequestBuilders.post(urlString, shortProblem.id)
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(challengeCreateRequestDtoString)
+                .header(HttpHeaders.AUTHORIZATION, "Bearer ${accessToken.token}")
+                .accept(MediaType.APPLICATION_JSON)
```
Comment (kshired) [inline src/test/kotlin/io/csbroker/apiserver/e2e/ProblemControllerV2Test.kt:172]: 🚫 **[ktlint]** <trailing-comma-on-call-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/test/kotlin/io/csbroker/apiserver/e2e/ProblemControllerV2Test.kt [hunk 7]
```diff
@@ -122,4 +137,66 @@ class ProblemControllerV2Test {
...
+        // then
+        result.andExpect(MockMvcResultMatchers.status().isOk)
+            .andExpect(MockMvcResultMatchers.content().string(CoreMatchers.containsString("success")))
+            .andDo(
+                document(
+                    "problems/v2/challenge",
+                    preprocessRequest(prettyPrint()),
+                    preprocessResponse(prettyPrint()),
+                    pathParameters(
+                        parameterWithName("problem_id").description("문제 id")
```
Comment (kshired) [inline src/test/kotlin/io/csbroker/apiserver/e2e/ProblemControllerV2Test.kt:184]: 🚫 **[ktlint]** <trailing-comma-on-call-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/test/kotlin/io/csbroker/apiserver/e2e/ProblemControllerV2Test.kt [hunk 8]
```diff
@@ -122,4 +137,66 @@ class ProblemControllerV2Test {
...
+                    ),
+                    requestFields(
+                        fieldWithPath("content").type(JsonFieldType.STRING)
+                            .description("이의제기 내용 ( 최소 10자 ~ 최대 150자 )")
```
Comment (kshired) [inline src/test/kotlin/io/csbroker/apiserver/e2e/ProblemControllerV2Test.kt:188]: 🚫 **[ktlint]** <trailing-comma-on-call-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/test/kotlin/io/csbroker/apiserver/e2e/ProblemControllerV2Test.kt [hunk 9]
```diff
@@ -122,4 +137,66 @@ class ProblemControllerV2Test {
...
+                    ),
+                    requestHeaders(
+                        headerWithName(HttpHeaders.AUTHORIZATION)
+                            .description("인증을 위한 Access 토큰")
+                            .optional()
```
Comment (kshired) [inline src/test/kotlin/io/csbroker/apiserver/e2e/ProblemControllerV2Test.kt:193]: 🚫 **[ktlint]** <trailing-comma-on-call-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/test/kotlin/io/csbroker/apiserver/e2e/ProblemControllerV2Test.kt [hunk 10]
```diff
@@ -122,4 +137,66 @@ class ProblemControllerV2Test {
...
+                    ),
+                    responseFields(
+                        fieldWithPath("status").type(JsonFieldType.STRING).description("결과 상태"),
+                        fieldWithPath("data").type(JsonFieldType.BOOLEAN).description("이의제기 데이터 생성 여부"),
+                    )
```
Comment (kshired) [inline src/test/kotlin/io/csbroker/apiserver/e2e/ProblemControllerV2Test.kt:198]: 🚫 **[ktlint]** <trailing-comma-on-call-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/test/kotlin/io/csbroker/apiserver/e2e/ProblemControllerV2Test.kt [hunk 11]
```diff
@@ -122,4 +137,66 @@ class ProblemControllerV2Test {
...
+                )
```
Comment (kshired) [inline src/test/kotlin/io/csbroker/apiserver/e2e/ProblemControllerV2Test.kt:199]: 🚫 **[ktlint]** <trailing-comma-on-call-site> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Missing trailing comma before ")"

File: src/main/kotlin/io/csbroker/apiserver/controller/v2/ProblemControllerV2.kt [hunk 12]
```diff
@@ -24,4 +30,15 @@ class ProblemControllerV2(
 
         return ApiResponse.success(findProblemDetail)
     }
+
+    @PostMapping("/{id}/challenge")
+    fun challengeProblem(
+        @LoginUser loginUser: User,
+        @PathVariable("id") id: Long,
+        @Valid
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/controller/v2/ProblemControllerV2.kt:38]: 🚫 **[ktlint]** <annotation> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Multiple annotations should not be placed on the same line as the annotated construct

File: src/main/kotlin/io/csbroker/apiserver/controller/v2/ProblemControllerV2.kt [hunk 13]
```diff
@@ -24,4 +30,15 @@ class ProblemControllerV2(
...
+        @RequestBody challengeCreateRequest: ChallengeCreateRequest,
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/controller/v2/ProblemControllerV2.kt:39]: 🚫 **[ktlint]** <annotation> <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>Annotation must be placed on separate line

File: src/main/kotlin/io/csbroker/apiserver/controller/v2/ProblemControllerV2.kt [hunk 14]
```diff
@@ -24,4 +30,16 @@ class ProblemControllerV2(
 
         return ApiResponse.success(findProblemDetail)
     }
+
+    @PostMapping("/{id}/challenge")
+    fun challengeProblem(
```
Comment (ekzm8523) [inline src/main/kotlin/io/csbroker/apiserver/controller/v2/ProblemControllerV2.kt:35]: challenge에 이의제기하다 라는 뜻이 있었군요..

File: src/main/kotlin/io/csbroker/apiserver/service/ProblemServiceImpl.kt [hunk 15]
```diff
@@ -533,4 +536,21 @@ class ProblemServiceImpl(
             GradeResultDto(gradingResponseDto)
         }
     }
+
+    @Transactional
+    override fun createChallenge(email: String, problemId: Long, content: String) {
```
Comment (ekzm8523) [inline src/main/kotlin/io/csbroker/apiserver/service/ProblemServiceImpl.kt:541]: 파라미터수가 조금 많은 것 같아요.. 줄일 수 있는 방법이 없을까요?

Comment (kshired) [reply src/main/kotlin/io/csbroker/apiserver/service/ProblemServiceImpl.kt]: @ekzm8523 done done dance~

File: src/main/kotlin/io/csbroker/apiserver/service/ProblemServiceImpl.kt [hunk 16]
```diff
@@ -538,7 +539,9 @@ class ProblemServiceImpl(
     }
 
     @Transactional
-    override fun createChallenge(email: String, problemId: Long, content: String) {
+    override fun createChallenge(createChallengeDto: CreateChallengeDto) {
+        val (email, problemId, content) = createChallengeDto
```
Comment (kshired) [inline src/main/kotlin/io/csbroker/apiserver/service/ProblemServiceImpl.kt:543]: 정말 별게 다 되는 Kotlin~

File: src/main/kotlin/io/csbroker/apiserver/service/ProblemServiceImpl.kt (hunk 15 above)
Comment (ekzm8523) [reply src/main/kotlin/io/csbroker/apiserver/service/ProblemServiceImpl.kt]: ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ 제발..

**Review by ekzm8523 — APPROVED** [review]
👍

**Pull request merged**